### PR TITLE
[Test][Cephadm] Add disconnected install

### PR DESCRIPTION
**Pull request by pranavprakash20**
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by pranavprakash20** [comment]
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-4QANS9/

**Comment by pranavprakash20** [comment]
successful bootstrap : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-F5VJ1T/

```
[root@ceph-prprakas-4w4f6x-node1-installer ~]# cephadm shell
Inferring fsid 92460c68-950a-11ee-81fb-fa163e560010
Using recent ceph image ceph-prprakas-4w4f6x-node1-installer:5000/rhceph/rhceph-5-rhel8@sha256:a193b0de114d19d2efd8750046b5d25da07e2c570e3c4eb4bd93e6de4b90a25a

[ceph: root@ceph-prprakas-4w4f6x-node1-installer /]# ceph -s
  cluster:
    id:     92460c68-950a-11ee-81fb-fa163e560010
    health: HEALTH_WARN
            Failed to place 1 daemon(s)
            1 failed cephadm daemon(s)
            OSD count 0 < osd_pool_default_size 3
 
  services:
    mon: 1 daemons, quorum ceph-prprakas-4w4f6x-node1-installer (age 2m)
    mgr: ceph-prprakas-4w4f6x-node1-installer.lkoexj(active, since 68s)
    osd: 0 osds: 0 up, 0 in
 
  data:
    pools:   0 pools, 0 pgs
    objects: 0 objects, 0 B
    usage:   0 B used, 0 B / 0 B avail
    pgs:     
 
```

**Comment by pranavprakash20** [comment]
Run log: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-7PXH77/

File: tests/ceph_installer/test_disconnected_install.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+PRIVATE_REG_USER = "myregistryusername"
+PRIVATE_REG_PASS = "myregistrypassword"
+PORT = "5000"
+
+
+class ClusterCustomContainerError(Exception):
+    pass
+
+
+def get_registry_details(config):
```
Comment (vamahaja) [inline tests/ceph_installer/test_disconnected_install.py:20]: instead use method - https://github.com/red-hat-storage/cephci/blob/master/cli/utilities/configs.py#L22

File: tests/ceph_installer/test_disconnected_install.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+log = Log(__name__)
+
+RHCEPH_5_REPO = "rhceph-5-tools-for-rhel-8-x86_64-rpms"
+PRIVATE_REG_USER = "myregistryusername"
+PRIVATE_REG_PASS = "myregistrypassword"
+PORT = "5000"
+
+
+class ClusterCustomContainerError(Exception):
```
Comment (vamahaja) [inline tests/ceph_installer/test_disconnected_install.py:16]: Unused exception

File: tests/ceph_installer/test_disconnected_install.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+from cephci.prereq import setup_private_container_registry
+from cephci.utils.configure import exec_cephadm_preflight, setup_ssh_keys
+from cli.exceptions import NotSupportedError
+from cli.utilities.packages import Package, SubscriptionManager
+from utility.log import Log
+from utility.utils import get_cephci_config
+
+log = Log(__name__)
```
Comment (vamahaja) [inline tests/ceph_installer/test_disconnected_install.py:8]: Unused variable

File: tests/ceph_installer/test_disconnected_install.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        private_reg_username=PRIVATE_REG_USER,
+        private_reg_password=PRIVATE_REG_PASS,
+        docker_reg_image=docker_reg_image,
+        images=[bootstrap_image],
+    )
+
+    # Perform bootstrap using the private registry created
+    mon_ip = mon_node.ip_address
+    registry_url = f"{mon_node.hostname}:{PORT}"
+    cmd = (
```
Comment (vamahaja) [inline tests/ceph_installer/test_disconnected_install.py:96]: instead use method - https://github.com/red-hat-storage/cephci/blob/master/cli/ops/cephadm.py#L125

File: tests/ceph_installer/test_disconnected_install.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        rhbuild = "reef"
+    elif rhbuild.startswith("6"):
+        rhbuild = "quincy"
+    elif rhbuild.startswith("5"):
+        rhbuild = "pacific"
+    else:
+        raise NotSupportedError(f"Unsupported version: {rhbuild}")
+
+    # Install cephadm and cephadm-ansible
+    SubscriptionManager(installer).repos.enable(RHCEPH_5_REPO)
```
Comment (vamahaja) [inline tests/ceph_installer/test_disconnected_install.py:56]: Why only rhceph5? shouldn't it depends on rhbuild ?

File: cephci/prereq.py [hunk 6]
```diff
@@ -191,7 +191,7 @@ def setup_private_container_registry(
         private_reg_username (str): private registry username
         private_reg_password (str): private registry password
         docker_reg_image (str): docker-registry image name
-        images (str): list of images
+        images (list): list of images
```
Comment (vamahaja) [inline cephci/prereq.py:194]: doesn't it affect existing tests ?

File: cephci/utils/configure.py [hunk 7]
```diff
@@ -314,7 +314,7 @@ def add_images_to_private_registry(
 
     for image in images:
         img_reg = image.replace(f"{registry}/", "")
-        src_image = f"{image}"
+        src_image = f"{registry}/{image}"
```
Comment (vamahaja) [inline cephci/utils/configure.py:317]: same as above

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cephci/prereq.py (hunk 6 above)
Comment (pranavprakash20) [reply cephci/prereq.py:194]: Its a typo fix Vaibhav, it was already accepting list of images. The description explains the same

File: cephci/utils/configure.py (hunk 7 above)
Comment (pranavprakash20) [reply cephci/utils/configure.py:317]: So looks like this was missed earlier, the ideal way as per 
https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/5/html-single/installation_guide/index#configuring-a-private-registry-for-a-disconnected-installation_install is as above Vaibhav

File: tests/ceph_installer/test_disconnected_install.py (hunk 3 above)
Comment (pranavprakash20) [reply tests/ceph_installer/test_disconnected_install.py:8]: Noted.

File: tests/ceph_installer/test_disconnected_install.py (hunk 1 above)
Comment (pranavprakash20) [reply tests/ceph_installer/test_disconnected_install.py:20]: Ack. Noted

File: tests/ceph_installer/test_disconnected_install.py (hunk 5 above)
Comment (pranavprakash20) [reply tests/ceph_installer/test_disconnected_install.py:56]: Made it generic. Now accepting it from config

**Comment by pranavprakash20** [comment]
logs: 
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-C0WVQ2/ (reason for updating cli/ops/cephadm.py )
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-X3AW4E/
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-C0OTJZ/

**Comment by pranavprakash20** [comment]
Also the upgrade might fail as the latest 7 is not updated in https://catalog.redhat.com/software/containers/search?product_listings_names=Red%20Hat%20Ceph%20Storage&p=1

**Review by MohitBis — APPROVED** [review]
LGTM

**Review by vamahaja — APPROVED** [review]

**Pull request merged**